PCHR-3000: Display Leave Request Actions for public holidays

reneolivo opened the pull request:
## Overview
This PR allows Admins to delete Public Holiday leave requests directly from the staff report/leave manager.

## Displaying the actions dropdown

### Before
![civihr_staff compucorp co uk hr17 9000 8](https://user-images.githubusercontent.com/1642119/33455390-48ea848c-d5f2-11e7-8b60-daa244219844.png)
The dropdown is missing

### After
![anim](https://user-images.githubusercontent.com/1642119/33455484-9898f914-d5f2-11e7-85da-13c172d0e16d.gif)

### Technical Details

To display the leave request actions dropdown, the `leave-request-actions` was inserted in the Public Holidays section of the staff leave report:

```html
<leave-request-actions
  leave-request="request"
  role="$root.section === 'my-leave' ? 'staff' : report.role"
  absence-types="report.absenceTypes"
  leave-request-statuses="report.leaveRequestStatuses">
</leave-request-actions>
```

To allow admins to see the Delete option for public holidays, the `leave-request-actions` was modified:

```js
function setAllowedActions () {
  // ...
  canDeletePublicHolidayLeaveRequests() && allowedActions.push('delete');
  // ....
}

function canDeletePublicHolidayLeaveRequests () {
  return vm.role === 'admin' && vm.leaveRequest.request_type === 'public_holiday';
}
```

## Moving cancelled requests to the "Cancelled and Others" section

A previous functionality ([this one](https://github.com/civicrm/civihr/blob/e9e203f351488ef16df4672dcc2de1a0302b9071/uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js#L477)) was removed from the code ([here](https://github.com/civicrm/civihr/pull/2017/files#diff-57dd767833c7cfbdf4f1f82ba6801faaL110)) where cancelled leave requests would be removed from their section and moved to the "Cancelled and Other" section.

This PR restores this removed functionality.

### Before
![anim](https://user-images.githubusercontent.com/1642119/33561711-c6975b54-d8e9-11e7-9c2c-179edbfb7728.gif)
The whole component refreshes so it can display the cancelled request

### After
![anim](https://user-images.githubusercontent.com/1642119/33561290-9ffb12d4-d8e8-11e7-8f6e-c730c2f4a853.gif)
The component is not refreshed and the functionality remains intact.

### Technical details

* the `LeaveRequest::deleted` event was renamed to `LeaveRequest::delete` so it follows other event and section name conventions (edit, new, delete vs deleted). This rename was done throughout all CiviHR projects.
* the `leave-request-actions.component.js`'s `publishEvents` method was refactored from this:

```js
function publishEvents (action) {
  // ...
  pubSub.publish('LeaveRequest::' + (action === 'delete' ? 'deleted' : 'edit'),
    vm.leaveRequest);
}
```

into this:

```js
function publishEvents (action) {
  // ...
  pubSub.publish('LeaveRequest::statusUpdate', {
    status: action,
    leaveRequest: vm.leaveRequest
  });
}
```

All components that listened for the `LeaveRequest::edit` now also listen for `LeaveRequest::statusUpdate`.

* `staff-leave-report.component.js` was made to follow the AngularJS styleguide regarding method definitions.
* `staff-leave-report.component.js`'s `registerEvents` method now listens for the `LeaveRequest::statusUpdate` event, and checks if the status is `cancel`, which removes the leave request from its current section and moves it to the cancel section:

```js
pubSub.subscribe('LeaveRequest::statusUpdate', handleStatusUpdate);

var statusUpdateHandlers = {
  delete: removeLeaveRequestFromItsSection,
  cancel: cancelLeaveRequestStatusHandler
};

function handleStatusUpdate (statusUpdate) {
  var handler = statusUpdateHandlers[statusUpdate.status];

  if (handler) {
    handler(statusUpdate.leaveRequest);
  } else {
    // fallbacks to component refresh is there are no handlers for the status provided:
    vm.refresh();
  }
}
```


### Refactoring the leave request removal

Deleting public holidays didn't trigger an update on the UI so the removal of these type of requests had to be supported. The removal of leave requests in general was refactored in `staff-leave-report.component.js` with the following changes:

1. the parameter and logic for `moveToOther` was removed from the `removeLeaveRequestFromItsSection` function since it was not being used at all.
2. instead of iterating only over `['approved', 'pending', 'other']`, the iteration occurs through all sections. For performance reasons the removal only occurs if the leave request is present in the section. To determine this, the requests are indexed by id after they are loaded:

```js
function callSectionLoadFn (section) {
  section.loading = true;

  return section.loadLeaveRequests()
    .then(indexSectionData.bind(this, section)) // <--- indexes the leave requests
    .then(function () {
      section.loading = false;
    });
}

function indexSectionData (section) {
  section.dataIndex = _.indexBy(section.data, 'id');
}
```

When removing a leave request, the function will loop through all sections, but will quickly determine if the leave request is part of the section or skip to the next:

```js
function removeLeaveRequestFromItsSection (leaveRequest) {
  _.forEach(vm.sections, function (section, sectionName) {
    var requestIndexInSection = section.dataIndex[leaveRequest.id];

    // not in section, so it skips the removal:
    if (!requestIndexInSection) {
      return;
    }

    // removes the leave request and the index:
    _.remove(section.data, function (dataEntry) {
      return dataEntry.id === leaveRequest.id;
    });
    delete section.dataIndex[leaveRequest.id];
    // recalculates the section numbers:
    updateSectionNumbersWithLeaveRequestBalanceChange(leaveRequest, sectionName);
  });
}
```

As a side note, there was a previous error with the section name for Public Holidays. Each absence type stored the section numbers for holidays, approved, and pending requests, but the "holidays" section was named "publicHolidays", so it was renamed to the real section name.

igorpavlov commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js:
```diff
@@ -121,18 +121,20 @@ define([
-      return section.loadFn().then(function () {
-        section.loading = false;
-      });
+      return section.loadFn()
```
If you refactored this anyway, may I ask you please to find a better name for the `loadFn` func?

igorpavlov commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js:
```diff
@@ -145,6 +147,16 @@ define([
+     * Indexes the leave request data of a section and stores it in the
+     * dataIndex attribute.
+     *
+     * @param {Object} section - the section object that contains data to index.
```
Could you please briefly describe the incoming param's structure?

igorpavlov commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js:
```diff
@@ -145,6 +147,16 @@ define([
+    }
+
     /**
      * NOTE: This should be just temporary, see PCHR-1810
```
Is this still relevant after refactoring? Not asking for any changes but would be great to know.

reneolivo replied:
Not sure why this is "temporary", looks like ok code to me. What should have been different? The ticket doesn't seem to be related to what this function accomplishes. Yes, they both are about leave request statuses, but this one is about loading all possible status values so they can be used to query leave request types, and the one in the ticket is about determining if a specific leave request has a determined status.

Lets wait for @deb1990 to see what kind of refactoring he had in mind.

igorpavlov commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js:
```diff
@@ -453,34 +465,24 @@ define([
-     *
-     * If the leave request belonged to either the "approved" or "pending" section,
-     * then the numbers of the section will be recalculated
+     * and the number of their sections are recalculated.
```
`Removes the given leave request from the section it currently belongs to and the number of their sections are recalculated` - not quite sure I understand the sentence. Could you please try to describe all in active voice?

reneolivo commented:
## New changes

### loadFn was renamed to loadLeaveRequests

[0791ab4](https://github.com/civicrm/civihr/pull/2344/commits/0791ab43d3b920e4d304fe0d121dfb00aeb3e154)

### Moving cancelled leave requests to the "Other" section

A previous functionality ([this one](https://github.com/civicrm/civihr/blob/e9e203f351488ef16df4672dcc2de1a0302b9071/uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/staff-leave-report.component.js#L477)) was removed from the code ([here](https://github.com/civicrm/civihr/pull/2017/files#diff-57dd767833c7cfbdf4f1f82ba6801faaL110)) where cancelled leave requests would be removed from their section and moved to the "Cancelled and Other" section.

This PR restores this functionality, and refactors the code. These are the changes done:

* the `LeaveRequest::deleted` event was renamed to `LeaveRequest::delete` so it follows other event and section name conventions (edit, new, delete vs deleted). This rename was done throughout all CiviHR projects.
* the `leave-request-actions.component.js`'s `publishEvents` method was refactored from this:

```js
function publishEvents (action) {
  // ...
  pubSub.publish('LeaveRequest::' + (action === 'delete' ? 'deleted' : 'edit'),
    vm.leaveRequest);
}
```

into this:

```js
function publishEvents (action) {
  var nonEditActions = ['delete', 'cancel'];
  // ...
  if (_.includes(nonEditActions, action)) {
    pubSub.publish('LeaveRequest::' + action, vm.leaveRequest);
  } else {
    pubSub.publish('LeaveRequest::edit', vm.leaveRequest);
  }
}
```

This allows for specific actions to be broadcasted (such as delete or cancel) and all other actions to fallback to the `LeaveRequest::edit` event.

* `staff-leave-report.component.js` was made to follow the AngularJS styleguide regarding method definitions.
* `staff-leave-report.component.js`'s `registerEvents` method now listens for the `LeaveRequest::cancel` event, which removes the leave request from its current section and moves it to the cancel section:

```js
pubSub.subscribe('LeaveRequest::cancel', function (leaveRequest) {
  removeLeaveRequestFromItsSection(leaveRequest);
  vm.sections.other.open && addLeaveRequestToSection(leaveRequest, vm.sections.other);
});
```
